File: ios/Tables/TableGridLayout.swift
import UIKit

enum TableLayoutDirection { case leftToRight, rightToLeft }

func tablePhysicalX(logicalX: CGFloat, width: CGFloat, totalWidth: CGFloat, rtl: Bool) -> CGFloat {
    rtl ? totalWidth - logicalX - width : logicalX
}

struct TableGridCell: Hashable {
    let sourcePosition: Int
    let row: Int
    let column: Int
    let rowspan: Int
    let colspan: Int
    let contentKey: String
    let attachmentRevision: Int

    init(sourcePosition: Int, row: Int, column: Int, rowspan: Int = 1, colspan: Int = 1,
         contentKey: String, attachmentRevision: Int = 0) {
        self.sourcePosition = sourcePosition; self.row = row; self.column = column
        self.rowspan = rowspan; self.colspan = colspan; self.contentKey = contentKey
        self.attachmentRevision = attachmentRevision
    }
}

struct TableGridRecord {
    let documentOwner: String
    let columns: Int
    let rows: Int
    let columnWidths: [CGFloat?]
    let cells: [TableGridCell]
    let failure: TableRenderFailure?
    let compatibilityDiagnostic: TableCompatibilityDiagnostic?

    init(documentOwner: String, columns: Int, rows: Int, columnWidths: [CGFloat?], cells: [TableGridCell],
         failure: TableRenderFailure? = nil, compatibilityDiagnostic: TableCompatibilityDiagnostic? = nil) {
        self.documentOwner = documentOwner; self.columns = columns; self.rows = rows
        self.columnWidths = columnWidths; self.cells = cells; self.failure = failure
        self.compatibilityDiagnostic = compatibilityDiagnostic
    }

    init(table: FfiViewerTable, documentOwner: String) {
        self.init(documentOwner: documentOwner, columns: Int(table.columns), rows: Int(table.rows),
                  columnWidths: table.columnWidths.map { $0.map(CGFloat.init) },
                  cells: table.cells.map { TableGridCell(sourcePosition: Int($0.sourcePos), row: Int($0.row), column: Int($0.column), rowspan: Int($0.rowspan), colspan: Int($0.colspan), contentKey: $0.contentKey) },
                  failure: table.failure, compatibilityDiagnostic: table.compatibilityDiagnostic)
    }
}

struct TableLayoutResult {
    let columnWidths: [CGFloat]
    let rowOffsets: [CGFloat]
    let rectangles: [Int: CGRect]
    let sourceOrder: [Int]
    let contentSize: CGSize
    let failure: TableRenderFailure?
    let compatibilityDiagnostic: TableCompatibilityDiagnostic?
}

final class TableGridLayout {
    private let displayScale: CGFloat
    private let cache: TableCellMeasurementCache

    init(displayScale: CGFloat = UIScreen.main.scale, cache: TableCellMeasurementCache = TableCellMeasurementCache()) {
        self.displayScale = displayScale.isFinite && displayScale > 0 ? displayScale : 1
        self.cache = cache
    }

    func layout(record: TableGridRecord, viewportWidth: CGFloat, style: TableStyle, direction: TableLayoutDirection,
                themeDigest: String = "", fontEnvironmentRevision: Int = 0, textScale: CGFloat = 1,
                measureCell: (TableGridCell, CGFloat) -> CGFloat?) -> TableLayoutResult {
        let minimumRow = style.cellPadding * 2 + style.borderWidth * 2
        let fallbackHeight = minimumRow.isFinite && minimumRow >= 1 ? minimumRow : 1
        let fallbackWidth: CGFloat
        if viewportWidth.isFinite, viewportWidth >= 0 {
            fallbackWidth = max(fallbackHeight, min(viewportWidth, style.minColumnWidth.isFinite ? style.minColumnWidth : fallbackHeight))
        } else {
            fallbackWidth = fallbackHeight
        }
        let invalidInput = !style.isValid || !viewportWidth.isFinite || viewportWidth < 0 || record.columns <= 0 || record.rows <= 0 || record.columnWidths.contains { ($0?.isFinite == false) || ($0 ?? 0) < 0 }
        guard !invalidInput, record.failure == nil else {
            return fallback(record.failure ?? .invalidAttributes, record, width: fallbackWidth, height: fallbackHeight)
        }
        var widths = (0..<record.columns).map { index -> CGFloat in
            let requested = index < record.columnWidths.count ? record.columnWidths[index] : nil
            return max(style.minColumnWidth, requested ?? 0)
        }
        let specified = widths.enumerated().filter { $0.offset < record.columnWidths.count && record.columnWidths[$0.offset] != nil }.map(\.offset)
        let unspecified = Set(0..<record.columns).subtracting(specified)
        let minimumWidth = widths.reduce(0, +)
        guard minimumWidth.isFinite else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
        let surplus = viewportWidth - minimumWidth
        if surplus > 0, !unspecified.isEmpty {
            let share = surplus / CGFloat(unspecified.count)
            for index in unspecified { widths[index] += share }
        }
        widths = widths.map { snapOutward($0) }
        guard widths.allSatisfy(\.isFinite) else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
        var xOffsets: [CGFloat] = [0]
        for width in widths {
            let offset = xOffsets.last! + width
            guard offset.isFinite else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
            xOffsets.append(offset)
        }
        var heights = Array(repeating: fallbackHeight, count: record.rows)
        let ordered = record.cells.sorted { $0.sourcePosition < $1.sourcePosition }
        guard ordered.allSatisfy({ valid($0, in: record) }) else {
            return fallback(.invalidStructure, record, width: fallbackWidth, height: fallbackHeight)
        }
        func contentHeight(_ cell: TableGridCell, _ inner: CGFloat) -> CGFloat? {
            let pixels = inner * displayScale
            let roundedPixels = pixels.rounded()
            guard roundedPixels.isFinite, roundedPixels >= 0,
                  let innerWidthPixels = Int(exactly: roundedPixels) else { return nil }
            let measuredWidth = CGFloat(innerWidthPixels) / displayScale
            let key = TableCellMeasurementKey(documentOwner: record.documentOwner, contentKey: cell.contentKey,
                                              innerWidthPixels: innerWidthPixels, themeDigest: themeDigest,
                                              fontEnvironmentRevision: fontEnvironmentRevision, textScale: textScale,
                                              attachmentRevision: cell.attachmentRevision)
            if let cached = cache.value(for: key) { return cached }
            guard let measured = measureCell(cell, measuredWidth), measured.isFinite, measured >= 0 else { return nil }
            let content = measured
            cache.insert(content, for: key)
            return content
        }
        for cell in ordered where cell.rowspan == 1 {
            let start = xOffsets[cell.column], end = xOffsets[cell.column + cell.colspan]
            let inner = max(0, end - start - 2 * (style.cellPadding + style.borderWidth))
            guard let content = contentHeight(cell, inner) else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
            let wanted = max(fallbackHeight, content + 2 * (style.cellPadding + style.borderWidth))
            guard wanted.isFinite else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
            heights[cell.row] = max(heights[cell.row], wanted)
        }
        for cell in ordered where cell.rowspan > 1 {
            let inner = max(0, xOffsets[cell.column + cell.colspan] - xOffsets[cell.column] - 2 * (style.cellPadding + style.borderWidth))
            guard let content = contentHeight(cell, inner) else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
            let wanted = max(fallbackHeight, content + 2 * (style.cellPadding + style.borderWidth))
            guard wanted.isFinite else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
            let covered = cell.row..<(cell.row + cell.rowspan)
            let current = covered.reduce(CGFloat(0)) { $0 + heights[$1] }
            guard current.isFinite else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
            if wanted > current {
                let height = heights[covered.upperBound - 1] + wanted - current
                guard height.isFinite else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
                heights[covered.upperBound - 1] = height
            }
        }
        var rows: [CGFloat] = [0]
        for height in heights {
            let offset = rows.last! + snapOutward(height)
            guard offset.isFinite else { return fallback(.invalidAttributes, record, width: fallbackWidth, height: fallbackHeight) }
            rows.append(offset)
        }
        let total = xOffsets.last!
        var rectangles: [Int: CGRect] = [:]
        for cell in ordered where valid(cell, in: record) {
            let logical = xOffsets[cell.column], width = xOffsets[cell.column + cell.colspan] - logical
            rectangles[cell.sourcePosition] = CGRect(x: tablePhysicalX(logicalX: logical, width: width, totalWidth: total, rtl: direction == .rightToLeft), y: rows[cell.row], width: width, height: rows[cell.row + cell.rowspan] - rows[cell.row])
        }
        return TableLayoutResult(columnWidths: widths, rowOffsets: rows, rectangles: rectangles, sourceOrder: ordered.map(\.sourcePosition), contentSize: CGSize(width: total, height: rows.last!), failure: nil, compatibilityDiagnostic: record.compatibilityDiagnostic)
    }

    private func valid(_ cell: TableGridCell, in record: TableGridRecord) -> Bool {
        cell.row >= 0 && cell.column >= 0 && cell.rowspan > 0 && cell.colspan > 0 &&
        cell.rowspan <= record.rows - cell.row && cell.colspan <= record.columns - cell.column
    }

    private func fallback(_ failure: TableRenderFailure, _ record: TableGridRecord, width: CGFloat, height: CGFloat) -> TableLayoutResult {
        TableLayoutResult(columnWidths: [], rowOffsets: [0, height], rectangles: [:], sourceOrder: [], contentSize: CGSize(width: width, height: height), failure: failure, compatibilityDiagnostic: record.compatibilityDiagnostic)
    }

    private func snapOutward(_ value: CGFloat) -> CGFloat { ceil(value * displayScale) / displayScale }
}

File: android/src/main/java/com/apollohg/editor/tables/TableGridLayout.kt
package com.apollohg.editor.tables

import uniffi.editor_core.FfiViewerTable
import uniffi.editor_core.TableCompatibilityDiagnostic
import uniffi.editor_core.TableRenderFailure
import kotlin.math.ceil

internal fun tablePhysicalX(logicalX: Float, width: Float, totalWidth: Float, rtl: Boolean): Float =
    if (rtl) totalWidth - logicalX - width else logicalX

data class TableGridCell(val sourcePosition: Int, val row: Int, val column: Int, val rowspan: Int = 1, val colspan: Int = 1, val contentKey: String, val attachmentRevision: Long = 0)

enum class TableLayoutFailure { GRID_LIMIT, WORK_LIMIT, ALLOCATION, INVALID_STRUCTURE, INVALID_ATTRIBUTES }

data class TableGridRecord(
    val documentOwner: String, val columns: Int, val rows: Int, val columnWidths: List<Float?>,
    val cells: List<TableGridCell>, val failure: TableLayoutFailure? = null,
    val compatibilityDiagnostic: TableCompatibilityDiagnostic? = null,
    val typedFailure: TableRenderFailure? = null
) {
    companion object {
        fun from(table: FfiViewerTable, documentOwner: String) = TableGridRecord(
            documentOwner, table.columns.toInt(), table.rows.toInt(), table.columnWidths.map { it?.toFloat() },
            table.cells.map { TableGridCell(it.sourcePos.toInt(), it.row.toInt(), it.column.toInt(), it.rowspan.toInt(), it.colspan.toInt(), it.contentKey) },
            table.failure?.let { TableLayoutFailure.valueOf(it.name) }, table.compatibilityDiagnostic, table.failure
        )
    }
}

data class TableCellRect(val left: Float, val top: Float, val width: Float, val height: Float)
data class TableLayoutResult(
    val columnWidths: List<Float>, val rowOffsets: List<Float>, val rectangles: Map<Int, TableCellRect>,
    val sourceOrder: List<Int>, val contentWidth: Float, val contentHeight: Float,
    val failure: TableLayoutFailure?, val compatibilityDiagnostic: TableCompatibilityDiagnostic?, val typedFailure: TableRenderFailure?
)

class TableGridLayout(private val displayScale: Float = 1f, private val cache: TableCellMeasurementCache = TableCellMeasurementCache()) {
    fun layout(record: TableGridRecord, viewportWidth: Float, style: TableStyle, rtl: Boolean,
               themeDigest: String = "", fontEnvironmentRevision: Long = 0, textScale: Float = 1f,
               measureCell: (TableGridCell, Float) -> Float?): TableLayoutResult {
        val minimumRow = style.cellPadding * 2f + style.borderWidth * 2f
        val fallbackHeight = if (minimumRow.isFinite() && minimumRow >= 1f) minimumRow else 1f
        val fallbackWidth = if (viewportWidth.isFinite() && viewportWidth >= 0f) {
            maxOf(fallbackHeight, minOf(viewportWidth, style.minColumnWidth.takeIf { it.isFinite() } ?: fallbackHeight))
        } else {
            fallbackHeight
        }
        val invalidInput = !style.isValid() || !viewportWidth.isFinite() || viewportWidth < 0f || record.columns <= 0 || record.rows <= 0 || record.columnWidths.any { it != null && (!it.isFinite() || it < 0f) }
        if (invalidInput || record.failure != null) {
            return fallback(record.failure ?: TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
        }
        val widths = MutableList(record.columns) { index -> maxOf(style.minColumnWidth, record.columnWidths.getOrNull(index) ?: 0f) }
        val unspecified = (0 until record.columns).filter { record.columnWidths.getOrNull(it) == null }
        val minimumWidth = widths.fold(0f) { total, width -> total + width }
        if (!minimumWidth.isFinite()) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
        val surplus = viewportWidth - minimumWidth
        if (surplus > 0f && unspecified.isNotEmpty()) unspecified.forEach { widths[it] += surplus / unspecified.size }
        for (index in widths.indices) widths[index] = snapOutward(widths[index])
        if (widths.any { !it.isFinite() }) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
        val xOffsets = mutableListOf(0f)
        for (width in widths) {
            val offset = xOffsets.last() + width
            if (!offset.isFinite()) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
            xOffsets += offset
        }
        val heights = MutableList(record.rows) { minimumRow }
        val ordered = record.cells.sortedBy { it.sourcePosition }
        if (!ordered.all { valid(it, record) }) {
            return fallback(TableLayoutFailure.INVALID_STRUCTURE, record, fallbackWidth, fallbackHeight)
        }
        fun contentHeight(cell: TableGridCell, inner: Float): Float? {
            val pixels = inner * scale()
            if (!pixels.isFinite() || pixels < 0f || pixels > Int.MAX_VALUE.toFloat()) return null
            val pixelWidth = kotlin.math.round(pixels).toInt()
            val measuredWidth = pixelWidth / scale()
            val key = TableCellMeasurementKey(record.documentOwner, cell.contentKey, pixelWidth, themeDigest, fontEnvironmentRevision, textScale, cell.attachmentRevision)
            cache.get(key)?.let { return it }
            val content = measureCell(cell, measuredWidth)?.takeIf { it.isFinite() && it >= 0f } ?: return null
            cache.put(key, content)
            return content
        }
        ordered.filter { it.rowspan == 1 }.forEach { cell ->
            val inner = maxOf(0f, xOffsets[cell.column + cell.colspan] - xOffsets[cell.column] - 2 * (style.cellPadding + style.borderWidth))
            val content = contentHeight(cell, inner) ?: return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
            val wanted = maxOf(fallbackHeight, content + 2 * (style.cellPadding + style.borderWidth))
            if (!wanted.isFinite()) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
            heights[cell.row] = maxOf(heights[cell.row], wanted)
        }
        ordered.filter { it.rowspan > 1 }.forEach { cell ->
            val inner = maxOf(0f, xOffsets[cell.column + cell.colspan] - xOffsets[cell.column] - 2 * (style.cellPadding + style.borderWidth))
            val content = contentHeight(cell, inner) ?: return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
            val wanted = maxOf(fallbackHeight, content + 2 * (style.cellPadding + style.borderWidth))
            if (!wanted.isFinite()) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
            val current = (cell.row until cell.row + cell.rowspan).sumOf { heights[it].toDouble() }.toFloat()
            if (!current.isFinite()) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
            if (wanted > current) {
                val height = heights[cell.row + cell.rowspan - 1] + wanted - current
                if (!height.isFinite()) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
                heights[cell.row + cell.rowspan - 1] = height
            }
        }
        val rows = mutableListOf(0f)
        for (height in heights) {
            val offset = rows.last() + snapOutward(height)
            if (!offset.isFinite()) return fallback(TableLayoutFailure.INVALID_ATTRIBUTES, record, fallbackWidth, fallbackHeight)
            rows += offset
        }
        val total = xOffsets.last()
        val rectangles = ordered.filter { valid(it, record) }.associate { cell ->
            val logical = xOffsets[cell.column]; val width = xOffsets[cell.column + cell.colspan] - logical
            cell.sourcePosition to TableCellRect(tablePhysicalX(logical, width, total, rtl), rows[cell.row], width, rows[cell.row + cell.rowspan] - rows[cell.row])
        }
        return TableLayoutResult(widths, rows, rectangles, ordered.map { it.sourcePosition }, total, rows.last(), null, record.compatibilityDiagnostic, null)
    }

    private fun fallback(failure: TableLayoutFailure, record: TableGridRecord, width: Float, height: Float) =
        TableLayoutResult(emptyList(), listOf(0f, height), emptyMap(), emptyList(), width, height, failure, record.compatibilityDiagnostic, record.typedFailure)

    private fun valid(cell: TableGridCell, record: TableGridRecord) = cell.row >= 0 && cell.column >= 0 && cell.rowspan > 0 && cell.colspan > 0 && cell.rowspan <= record.rows - cell.row && cell.colspan <= record.columns - cell.column
    private fun scale() = if (displayScale.isFinite() && displayScale > 0f) displayScale else 1f
    private fun snapOutward(value: Float) = ceil(value * scale()) / scale()
}
